stdlib/SparseArrays/src/linalg.jl

```diff
@@ -1,12 +1,44 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
-import LinearAlgebra: checksquare
+import LinearAlgebra: checksquare, sym_uplo
 using Random: rand!
 
 # In matrix-vector multiplication, the correct orientation of the vector is assumed.
 const StridedOrTriangularMatrix{T} = Union{StridedMatrix{T}, LowerTriangular{T}, UnitLowerTriangular{T}, UpperTriangular{T}, UnitUpperTriangular{T}}
 const AdjOrTransStridedOrTriangularMatrix{T} = Union{StridedOrTriangularMatrix{T},Adjoint{<:Any,<:StridedOrTriangularMatrix{T}},Transpose{<:Any,<:StridedOrTriangularMatrix{T}}}
 
+for HermSym ∈ [:Hermitian, :Symmetric]
+    @eval begin
+        function (+)(A::SparseMatrixCSC, B::$HermSym{<:Any,<:AbstractSparseMatrix})
+            C  = sparse(B)
+            C .= A .+ C
+        end
+        function (+)(A::SparseMatrixCSC, B::$HermSym)
+            C  = collect(B)
+            C .= A .+ C
+        end
+        (+)(A::$HermSym{<:Any,<:AbstractSparseMatrix}, B::SparseMatrixCSC) = B + A
+        (+)(A::$HermSym, B::SparseMatrixCSC) = B + A
+
+        function +(A::$HermSym{<:Any, <:SparseMatrixCSC}, B::$HermSym{<:Any, <:SparseMatrixCSC})
+            if A.uplo == B.uplo
+                $HermSym(parent(A) + parent(B), sym_uplo(A.uplo))
+            elseif A.uplo == 'U'
+                $HermSym(parent(A) + permutedims(parent(B)), :U)
+            else
+                $HermSym(permutedims(parent(A)) + parent(B), :U)
```

Comment:
I think it's Julia style to have a `return` in each line, because we are not exiting the function in some unique final line.

Reply:
This function causes an ambiguity error with

https://github.com/JuliaLang/julia/blob/0c284839fef6c8c153edc01fddfa37a9f5ac6752/stdlib/LinearAlgebra/src/symmetric.jl#L465-L466

so I guess we should tweak the version there to get the desired behavior. In fact, I've played with it a little bit, and making the above code the generic one seems greatly beneficial:
```julia
using SparseArrays, LinearAlgebra, Test
import Base: +
import LinearAlgebra: sym_uplo
using BenchmarkTools

function myplus(A::Hermitian, B::Hermitian)
    if A.uplo == B.uplo
        Hermitian(parent(A) + parent(B), sym_uplo(A.uplo))
    elseif A.uplo == 'U'
        Hermitian(parent(A) + permutedims(parent(B)), :U)
    else
        Hermitian(permutedims(parent(A)) + parent(B), :U)
    end
end

A = sprandn(100, 100, 0.1); B = sprandn(100, 100, 0.1);

@btime $(Hermitian(A)) + $(Hermitian(B));
@btime myplus($(Hermitian(A)), $(Hermitian(B)));
@test Hermitian(A) + Hermitian(B) ≈ myplus(Hermitian(A), Hermitian(B))
Ad, Bd = Matrix.((A, B));
@btime $(Hermitian(Ad)) + $(Hermitian(Bd));
@btime myplus($(Hermitian(Ad)), $(Hermitian(Bd)));
@test Hermitian(Ad) + Hermitian(Bd) ≈ myplus(Hermitian(Ad), Hermitian(Bd))
```
I think this is because we're working with the underlying array directly, instead of calling `getindex` on a `Hermitian` many times, as happens in the above-quoted code.